website/blog/2023-02-14-pgo.md

```diff
@@ -0,0 +1,101 @@
+---
+title: Optimizing Databend Binary Builds with Profile-guided Optimization
+description: rustc supports doing profile-guided optimization (PGO). Let's see how we can make it work for Databend.
+slug: profile-guided-optimization
+date: 2023-02-13
+tags: [databend, Profile-guided Optimization]
+authors:
+- name: PsiACE
+  url: https://github.com/PsiACE
+  image_url: https://github.com/PsiACE.png
+---
+
+Previously, a voice in the community suggested that we try profile-guided optimization ([#9387](https://github.com/datafuselabs/databend/issues/9387)). Let's see how we can use Rust to build a PGO-optimised Databend!
+
+## Background
+
+Profile-guided Optimization is a compiler optimisation technique, which collects typical execution data (possible branches) during program execution and then optimises for inlining, conditional branches, machine code layout, register allocation, etc.
+
+The background to the introduction of this technique is that static analysis techniques only consider code performance improvements without actually executing the program. However, these optimisations may not be fully effective. In the absence of runtime information, the compiler cannot take into account the actual execution of the program.
+
+PGO allows data to be collected based on application scenarios in a production environment, allowing the optimiser to optimise speed for hot code paths and size for cold code paths, producing faster and smaller code for applications.
+
+rustc supports PGO, with data collection built into the binaries and then perf data collected during runtime in preparation for final compilation optimisation. The implementation relies entirely on LLVM.
+
+## Workflow
+
+Generating a PGO-optimized program involves following a workflow with four steps:
+
+- Compile the program with instrumentation enabled.
+- Run the instrumented program which generates a `profraw` file.
+- Convert the `.profraw` file into a `.profdata` file using LLVM's `llvm-profdata` tool.
+- Compile the program again, this time making use of the profiling data.
+
+## Preparation
+
+The data collected during the run will eventually need to be converted with `llvm-profdata`, which can be provided by installing the `llvm-tools-preview` component via `rustup`, or consider using recent versions of LLVM and the program provided by Clang.
+
+```bash
+rustup component add llvm-tools-preview
+```
+
+After installation `llvm-profdata` may need to be added to the `PATH` at the following path.
+
+```bash
+~/.rustup/toolchains/<toolchain>/lib/rustlib/<target-triple>/bin/
+```
+
+## Step-By-Step
+
+Rather than choosing a specific production environment workload, Databend's SQL logic tests are used here as a demonstration. It may not be positive in performance, but it can help us to understand how the process is carried out.
```

Comment:
Rather than choosing a specific production environment workload, Databend's SQL logic tests are used here as a demonstration.  ----------I dont understand here.

Reply:
but it can help us to understand ...-> but it can help us understand ...